src/graphics/text.rs

```diff
@@ -346,6 +349,169 @@ impl Drawable for Text {
+        TextLayout {
+            font_atlas: font_atlas,
+            chars: Vec::new(),
+            words: Vec::new(),
+            cursor: Point2::new(0., 0.),
+            space_width: space_width,
+        }
+    }
+
+    pub fn push_str(&mut self, text: &str, colors: impl IntoIterator<Item = Color> + Clone) {
+        if let Some(upper_bound) = colors.clone().into_iter().size_hint().1 {
+            assert!(
+                upper_bound < text.len(),
+                "Passed in less colors than the number of chars you tried to push!"
+            );
+        }
+        self.words.append(&mut Word::from_str(
+            text,
+            &self.font_atlas.font_map,
+            self.words.last().unwrap_or(&Word { end: 0, width: 0. }).end,
+        ));
+        for (c, color) in text.chars().zip(colors.clone()) {
+            if c.is_whitespace() {
+                self.cursor.x += self.space_width;
+                continue;
+            }
+            let c_info = self
+                .font_atlas
+                .font_map
+                .get(&c)
+                .unwrap_or(&self.font_atlas.font_map[&'?']);
+            self.chars.push(LayoutCharInfo {
+                coords: Box2::new(
+                    self.cursor.x + c_info.horizontal_offset,
+                    self.cursor.y + c_info.vertical_offset,
+                    c_info.width,
+                    c_info.height,
+                ),
+                color,
+                c,
+            });
+            self.cursor.x += c_info.advance_width;
+        }
+    }
+
+    pub fn push_wrapping_str(
+        &mut self,
+        text: &str,
+        colors: impl IntoIterator<Item = Color> + Clone,
+        line_width: f32,
+    ) {
+        if let Some(upper_bound) = colors.clone().into_iter().size_hint().1 {
+            assert!(
+                upper_bound < text.len(),
+                "Passed in less colors than the number of chars you tried to push!"
+            );
+        }
+
+        let new_words = Word::from_str(
+            text,
+            &self.font_atlas.font_map,
+            self.words.last().unwrap_or(&Word { end: 0, width: 0. }).end,
+        );
+
+        let mut start = match self.words.last() {
+            Some(w) => w.end,
+            None => 0usize,
+        };
+
+        let mut char_iter = text.chars();
+        let mut color_iter = colors.into_iter();
+
+        for word in new_words.iter() {
+            if word.width + self.cursor.x > line_width as f32 {
+                self.cursor.x = 0.;
+                self.cursor.y += self.font_atlas.line_gap;
+            }
+
+            for _ in 0..(word.end - start) {
+                let c = char_iter.next().unwrap();
+                let color = color_iter.next().unwrap();
+                let c_info = self
+                    .font_atlas
+                    .font_map
+                    .get(&c)
+                    .unwrap_or(&self.font_atlas.font_map[&'?']);
```

Comment:
you don't need to index this every time. why not grab `&self.font_atlas.font_map[&'?']` outside the loop and hold onto it? it's a hashmap access so Rust can't really all that easily optimize it into a variable access and hoist it out of the loops